gio/src/cancellable.rs

```diff
@@ -1,10 +1,238 @@
+    #[inline]
+    unsafe fn try_from_glib(val: libc::c_ulong) -> Result<Self, GlibNoneError> {
+        NonZeroU64::new(val as u64).map(Self).ok_or(GlibNoneError)
+    }
+}
+
+pub trait CancellableExtManual {
+    // rustdoc-stripper-ignore-next
+    /// Convenience function to connect to the `signal::Cancellable::cancelled` signal. Also
+    /// handles the race condition that may happen if the cancellable is cancelled right before
+    /// connecting. If the operation is cancelled from another thread, `callback` will be called
+    /// in the thread that cancelled the operation, not the thread that is running the operation.
+    /// This may be the main thread, so the callback should not do something that can block.
+    ///
+    /// `callback` is called at most once, either directly at the time of the connect if `self` is
+    /// already cancelled, or when `self` is cancelled in some thread.
+    ///
+    /// Since GLib 2.40, the lock protecting `self` is not held when `callback` is invoked. This
+    /// lifts a restriction in place for earlier GLib versions which now makes it easier to write
+    /// cleanup code that unconditionally invokes e.g.
+    /// [`CancellableExt::cancel()`][crate::prelude::CancellableExt::cancel()].
+    ///
+    /// # Returns
+    ///
+    /// The id of the signal handler or `None` if `self` has already been cancelled.
+    #[doc(alias = "g_cancellable_connect")]
+    fn connect_cancelled<F: FnOnce(&Self) + Send + Sync + 'static>(
+        &self,
+        callback: F,
+    ) -> Option<CancelledHandlerId>;
+    // rustdoc-stripper-ignore-next
+    /// Same as [`Self::connect_cancelled`] but takes a non-`Send+Sync` closure. If the cancellable
+    /// is cancelled from a different thread than it was connected to then the cancellation will
+    /// panic.
+    fn connect_cancelled_local<F: FnOnce(&Self) + 'static>(
+        &self,
+        callback: F,
+    ) -> Option<CancelledHandlerId>;
+    // rustdoc-stripper-ignore-next
+    /// Same as [`Self::connect_cancelled`] but takes a non-`Send+Sync` and non-`'static'` closure.
+    /// No runtime checks are performed for ensuring that the closure is called correctly.
+    unsafe fn connect_cancelled_unsafe<F: FnOnce(&Self)>(
+        &self,
+        callback: F,
+    ) -> Option<CancelledHandlerId>;
+    // rustdoc-stripper-ignore-next
+    /// Disconnects a handler from a cancellable instance. Additionally, in the event that a signal
+    /// handler is currently running, this call will block until the handler has finished. Calling
+    /// this function from a callback registered with [`Self::connect_cancelled`] will therefore
+    /// result in a deadlock.
+    ///
+    /// This avoids a race condition where a thread cancels at the same time as the cancellable
+    /// operation is finished and the signal handler is removed.
+    #[doc(alias = "g_cancellable_disconnect")]
+    fn disconnect_cancelled(&self, id: Option<CancelledHandlerId>);
```

Comment:
```suggestion
    fn disconnect_cancelled(&self, id: CancelledHandlerId);
```

Why would you ever want to disconnect a `None`?

Reply:
`g_cancellable_disconnect` lets you pass 0 so you don't have to write the if statement, but it's probably better in Rust to require the if statement to avoid an unnecessary FFI boundary